---
title: "Strings and Functions"
author: 
output: html_document
---


```{r}
library (stringr)
library(readr)
codons = read_csv("codons")
```

setwd("/Users/Lamb/Documents/GitHub/Stats-133/hw2")
getwd()

## Introduction

This assignment has a few purposes:

* to familiarize you with manipulating character strings
* to introduce you to regular expressions in R
* to get you to practice reading documentation of R packages
* to get you thinking about the structure of functions
* to get you to organize your GitHub repository

## Submission Notes

* Remember not to overwrite my file! You should immediately save the Rmd file and the txt file for problem 6 in your own stat133 directory
* Submit the `html` and `rmd` file onto bCourses
* Place the two files into the appropriate folder in your GitHub repository (see below for organization)
* Push to GitHub
* You'll share the repo with us in assignment 3

## Allowed Packages
You are allowed to load the following packages for this assignment:

* `stringr`
* `readr`

## File Organization (GitHub Points)
Your GitHub repository should contain 18 folders:

* 1 folder for each of the 8 homework assignments labelled `hw1`, `hw2`, ..., `hw8`
* 1 folder for each of the 9 weeks of labs labelled
`labs1`, `labs2`, ..., `labs9`
* 1 folder labelled `final_project`

Place your homework and lab files in the appropriate folders.  Lab 1 and 3 should be in `labs1`.  Labs 5-7 should be in `labs2`.

## Problems 

### Problem 1 - Names of Files (8 pts)

Imagine that you need to generate the names of 4 data files (with .png extension). All the files have the same prefix name but each of them has a different number: `plot001.png`, `plot002.png`, `plot003.png`, and `plot004.png`. We can generate a character vector with these names in R. One naive solution would be to write something like this:
```{r}
files <- c('plot001.png', 
           'plot002.png', 
           'plot003.png', 
           'plot004.png')
```

Now imagine that you need to generate 100 file names. You could write a vector with 100 file names but it's going to take you a while.

Generate a character vector named `files100` containing 100 files names of the form `plotxxx.png` i.e. the names should look like `plot001.png`, `plot002.png`, `plot003.png`, ..., `plot099.png`, `plot100.png`.

Do not use any looping constructions (we haven't covered this yet, and they're slower in `R` for this task).

```{r}

firstnum = paste("plot", "00", seq(1,9), ".png", sep = "")
secondnum = paste("plot", "0", seq(10,99), ".png", sep ="")
lastnum = ("plot100.png")

files100 = c(firstnum, secondnum, lastnum)
  
files100


```

### Problem 2 - USA States Names (8 pts)

library(stringr)

One of the datasets that comes with R is `USArrests`. The row names of this data set correspond to the 50 states. This next problem refers to the `states` vector below.  Ignore upper and lower case.

```{r}

states <- rownames(USArrests)

```

1. How many states contain the letter `i`?

```{r}

length(str_subset(states, "i"))

```



2. How many states do not contain the letter `a`?
```{r}

length(states) - length(str_subset(states, "[a]"))

```

3. Which states contain the letter `j`?
```{r}

str_subset(states, regex("j", ignore_case = TRUE))


```

4. Which states are formed by two words?
```{r}

str_subset(states, " ")


```

5. Which states start with `W` and end with a consonant (non-vowel)?
```{r}


str_subset(states, "^[W].*[^aeiou]$")

```

6. Which states contain at least three `i`'s (e.g. Illinois)?
```{r}


str_subset(states, regex("([i].*){3,}", ignore_case = TRUE))

```

7. Which states contain exactly five vowels (e.g. California)?
```{r}

str_subset(states,"([AEIOUaeiou].*){5}")


```

8. Which states contain three consecutive vowels (e.g. Hawaii)?
```{r}

str_subset(states, "[AEIOUaeiou]{3}")

```

### Problem 3 Colors in Hexadecimal Notation (4 pts)

Write a function `is_hex()` that checks whether the input is a valid color in hexadecimal notation. A hex color starts with a hash `#` symbol followed by six hexadecimal digits: 0 to 9, and the first six letters A, B, C, D, E, F. Your function should not distinguish between upper and lower case letters.

```{r}

is_hex = function(x)

{
  

str_detect(x, "^[#]([0-9ABCDEFabcdef]){6}") & str_length(x) == 7
  
}

```

Change `eval` to `TRUE` to test your function.
```{r, eval=TRUE}
is_hex("#ff0000")  # TRUE
is_hex("#123456")  # TRUE
is_hex("#12Fb56")  # TRUE

is_hex("#1234GF")  # FALSE
is_hex("#1234567") # FALSE
is_hex("blue")     # FALSE
```

### Problem 4 - Reversing Characters (4 pts)

Write a function called `reverse_chars` that reverses a string by characters. You may find the `rev` function useful.

```{r}

reverse_chars = function(x)
  {
  
  sequence = str_sub(x, seq(1,nchar(x)), seq(1,nchar(x)))

  return(str_c(rev(sequence), collapse = ""))

}

```

Change `eval` to `TRUE` to test your function.
```{r, eval=TRUE}
reverse_chars("Rictusempra")
reverse_chars("Lumox Maxima")
reverse_chars("Repello Muggletum")
```

### Problem 5 - Cleaning XML Data (4 points)

Rarely does text data come clean enough for immediate use.  In this exercise, you'll get a taste of parsing XML Data, which a non-tabular data format that makes use of tags.  The data are sometimes presented in the following form:
```
<tag_name>data</tag_name>
```
Your job is to write a function `remove_tag` that takes a vector of text from an XML document and removes the tags, returning a character vector with only the data values.  So, for example:

* `<common>dandelion</common>` is converted to `dandelion`
* `<genus>Taraxacum</genus>` is converted to `Taraxacum`
* `<kingdom>Plantae</kingdom>` is converted to `Plantae`

```{r}

remove_tag = function(x)
{ 
  
  str_replace_all(x,"<.*?>", "")

}

```

Test your code:
```{r, eval = TRUE}
phylo <- c("<common>dandelion</common>",
           "<genus>Taraxacum</genus>",
           "<kingdom>Plantae</kingdom>")
remove_tag(phylo)
```

### Problem 6 - Processing DNA (12 points)

Our genetic blueprint, DNA, consists of nucleotides each composed of a sugar, deoxyribose; a phosphate group; and one of four nitrogen-containing nucleobases: cytosine (C), guanine (G), adenine (A), and thymine (T).  Simplifying the biology a little (ok, a lot), every three bases in a corresponds to an amino acid, a building block for proteins.  For example, the triplets `TTT-CTT-ATT` corresponds to the amino acids _Phenylalanine_, _Leucine_, and _Isoleucine_, which is then abbreviated as `F-L-I`.  

That's the extent of the biology you need to know, but if you want to read more of the technicalities, [see this wikipedia article for more information.](https://en.wikipedia.org/wiki/DNA_codon_table)  We ignore START codons in this problem.

A csv file of the DNA to amino acid conversion is given with the homework.  Use a relative path from your `hw2` directory to load that data frame in with the `read_csv` function in the `readr` package.

Write a function `dna_to_aa` that converts a sequence of DNA given as a string into a sequence of amino acids.  If the sequence contains an invalid DNA character (something outside CGAT), your function should return an error stating "Invalid DNA sequence".


Hints: 

* Break this problem down into steps.  This is more than a one-liner
* `str_replace_all` can take a named vector as one of its arguments.

```{r}

dna_to_aa = function(x)
{
  detector = str_detect(x, "[^CGAT]")
  
  if (detector == TRUE) 
  {
    return ("Invalid DNA sequence")
  }
  
  else {
    
    sections = substring(x, seq(1, nchar(x), 3), seq(3, nchar(x), 3))
  
  
    great = match(sections, codons$dna_triplet)
    

    
    return(paste(str_replace_all(sections, "[GTAC]{3}", codons$amino_acid[great]), sep="", collapse = ""))
    
   
    }
   
  
  }
  


```

Test your function:
```{r, eval = T}
dna_to_aa("GAGCCATGCATT") # "EPCI"
dna_to_aa("ATCTAGATAGTAGGC") # "IStopIVG"
dna_to_aa("UCUGAGAAU") # "Invalid DNA sequence"
```




